apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,292 @@
+
+### `sampled_addmm` (SDD)
+In the case of this operation a dense equivalent would involve the composition between matrix
+multiplication and a masking operation.  Here, we have chose for the dense baseline a normal matrix
+multiplication without masking. Since in practice the masking operation may not be used if sparse
+tensors are not also in use, for example when computing gradients. This gives us a harder target to
+hit, but is a fair assessment. All matrices are square two with size 4096 in both dimensions, this
+is a large subject, but sparsity is usually attractive in these cases as one of the features is a
+reduction in memory footprint.
+
+A key detail on these figures is the sparsity ratio (horizontal axis) where the speedup (vertical
+axis) crosses the 1.0 threshold, which is marked with a horizontal line labeled "1x". At this point
+the sparse operation executes more quickly than the dense baseline. This quantity tells us how much
+do we need to be able to prune to use sparsity without our model executing more slowly. Looking at
+the results for [`float32`][#fig-2] data type,
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/sampled_addmm_float32.png"
+    name="fig-2"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the dense baseline on the vertical.  There is a horizontal line marking the speedup ratio of 1.0
+    above which the sparse subject takes less time to execute than the dense counterpart.  Three lines are
+    plotted indicating block sizes used for the sparse tensor of 16, 32, and 64."
+/>
+
+we see some promising results. Using block sizes of 32 and 64, we cross the 1x threshold below 70%
+sparsity. Especially promising is that the smallest block size of 16 is able to reach 1x below 90%
+sparsity. The small block size is important as it offers more flexibility in the pruning stage as
+non-zero values can be more dispersed usually resulting in a less accuracy loss.
+
+Moving on to [half precision data types][#fig-3] we see much poorer performance compared to dense.
+With the 1x threshold crossed only be the largest block size (64) and only a sparsity greater than
+90%. Requiring pruning to this degree is not practical, in most cases pruning to this level will
+result in unacceptable drops in accuracy. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/sampled_addmm_half.png"
+    name="fig-3"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the dense baseline on the vertical.  There is a horizontal line marking the speedup ratio of 1.0
+    above which the sparse subject takes less time to execute than the dense counterpart.  Three lines are
+    plotted indicating block sizes used for the sparse tensor of 16, 32, and 64."
+/>
+
+We clearly have more work to do improve the performance of `sampled_addmm`, but our first pass still
```

Comment:
Don't be so harsh on yourselves. Rewrite this in a more impersonal way. Something like "These results, albeit promising, show that there is still room for improvement"